packages/next/server/base-http.ts

```diff
@@ -0,0 +1,287 @@
+import type { ServerResponse, IncomingMessage, IncomingHttpHeaders } from 'http'
+import type { Writable, Readable } from 'stream'
+import { PERMANENT_REDIRECT_STATUS } from '../shared/lib/constants'
+import {
+  getCookieParser,
+  NextApiRequestCookies,
+  parseBody,
+  SYMBOL_CLEARED_COOKIES,
+} from './api-utils'
+import { I18NConfig } from './config-shared'
+import { NEXT_REQUEST_META, RequestMeta } from './request-meta'
+
+export interface BaseNextRequestConfig {
+  basePath: string | undefined
+  i18n?: I18NConfig
+  trailingSlash?: boolean | undefined
+}
+
+export abstract class BaseNextRequest<Body = any> {
+  protected _cookies: NextApiRequestCookies | undefined
+  public abstract headers: IncomingHttpHeaders
+
+  constructor(public method: string, public url: string, public body: Body) {}
+
+  abstract parseBody(limit: string | number): Promise<any>
+
+  // Utils implemented using the abstract methods above
+
+  public get cookies() {
+    if (this._cookies) return this._cookies
+    return (this._cookies = getCookieParser(this.headers)())
+  }
+}
+
+export class NodeNextRequest extends BaseNextRequest<Readable> {
+  public headers = this._req.headers;
+
+  [NEXT_REQUEST_META]: RequestMeta
+
+  get req() {
```

Comment:
I think we should unify both `NodeNextRequest` and `WebNextRequest` to have `.originalRequest` as the original request instance. Here we have `.req` for the Node one and `.request` for the Web one.